impl/src/main/java/org/jboss/weld/xml/BeansXmlStreamParser.java

```diff
@@ -0,0 +1,398 @@
+ * Simple yet efficient parser for beans.xml. This class is not thread safe and instances cannot be reused.
+ *
+ * @author Martin Kouba
+ */
+public class BeansXmlStreamParser {
+
+    public static final String JAVAEE_LEGACY_URI = "http://java.sun.com/xml/ns/javaee";
+    public static final String JAVAEE_URI = "http://xmlns.jcp.org/xml/ns/javaee";
+    public static final Set<String> JAVAEE_URIS = ImmutableSet.of(JAVAEE_LEGACY_URI, JAVAEE_URI);
+
+    public static final String WELD_URI = "http://jboss.org/schema/weld/beans";
+    public static final Set<String> SCANNING_URIS = ImmutableSet.of(WELD_URI, JAVAEE_URI, JAVAEE_LEGACY_URI);
+
+    private static final String VERSION_ATTRIBUTE_NAME = "version";
+    private static final String BEAN_DISCOVERY_MODE_ATTRIBUTE_NAME = "bean-discovery-mode";
+    private static final String NAME_ATTRIBUTE_NAME = "name";
+    private static final String VALUE_ATTRIBUTE_NAME = "value";
+    private static final String PATTERN_ATTRIBUTE_NAME = "pattern";
+
+    private static final String IF_CLASS_AVAILABLE = "if-class-available";
+    private static final String IF_CLASS_NOT_AVAILABLE = "if-class-not-available";
+    private static final String IF_SYSTEM_PROPERTY = "if-system-property";
+
+    private static final String CLASS = "class";
+    private static final String STEREOTYPE = "stereotype";
+    private static final String INCLUDE = "include";
+    private static final String EXCLUDE = "exclude";
+    private static final String TRIM = "trim";
+    private static final String BEANS = "beans";
+    private static final String ALTERNATIVES = "alternatives";
+    private static final String INTERCEPTORS = "interceptors";
+    private static final String DECORATORS = "decorators";
+    private static final String SCAN = "scan";
+
+    private List<Metadata<String>> enabledInterceptors = null;
+    private List<Metadata<String>> enabledDecorators = null;
+    private List<Metadata<String>> selectedAlternatives = null;
+    private List<Metadata<String>> selectedAlternativeStereotypes = null;
+    private List<Metadata<Filter>> includes = null;
+    private List<Metadata<Filter>> excludes = null;
+    private BeanDiscoveryMode discoveryMode = BeanDiscoveryMode.ALL;
+    private String version;
+    private boolean isTrimmed;
+
+    private final URL beansXml;
+
+    private final Function<String, String> interpolator;
+
+    /**
+     *
+     * @param beansXml
+     */
+    public BeansXmlStreamParser(URL beansXml) {
+        this(beansXml, Function.identity());
+    }
+
+    /**
+     *
+     * @param beansXml
+     * @param interpolator
+     */
+    public BeansXmlStreamParser(URL beansXml, Function<String, String> interpolator) {
+        this.beansXml = beansXml;
+        this.interpolator = interpolator;
+    }
+
+    public BeansXml parse() {
+        if (beansXml == null) {
+            throw XmlLogger.LOG.loadError("unknown", null);
+        }
+        try (InputStream in = beansXml.openStream()) {
```

Comment:
Maybe buffer this?

Reply:
I'd leave this to `XMLInputFactory` implementation...

Reply:
Minor point, but I disagree; the "creator" of an input stream determines how or if it's buffered.  The "wrapper" (the `XMLInputFactory` in this case) normally just uses it as supplied.

Reply:
Except the contract is not defined. `beans.xml` are usually small so I don't think it's worth the effort.